### PR TITLE
Move snapshot APIs and client to v1 (no controller changes)

### PR DESCRIPTION
**Pull request by xing-yang**
**What type of PR is this?**
/kind api-change

**What this PR does / why we need it**:
This PR moves snapshot APIs and client to v1.  This only includes API and client changes.
Controller changes are in a separate PR.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
Move Snapshot APIs to V1.
```

### CODE REVIEW

**Comment by xing-yang** [comment]
Hi @yuxiangqian,
Michelle suggested that I submit one PR including only API and clientset changes so we can pass CI.  After that submit another PR with controller changes.

Here is the PR including only API and clientset changes (same as https://github.com/kubernetes-csi/external-snapshotter/pull/389 except that it does not have controller changes).  Please take a look.  Thanks.

File: client/apis/volumesnapshot/v1/types.go
Comment (yuxiangqian) [inline client/apis/volumesnapshot/v1/types.go:2]: 2020?

File: client/hack/README.md
Comment (yuxiangqian) [inline client/hack/README.md:102]: the "oneOf" sections in v1 is manually added, will those be removed when run controller-gen? if so, we will need to document that.

**Review by yuxiangqian — COMMENTED** [review]
overall lgtm with some nits. It's really hard to review this type of changes :). I'd highly suggest, if there is next time,  to only include those files which are not auto-generated in the first PR (even though the build/test might fail) to ease the process.

File: client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml
Comment (xing-yang) [reply client/config/crd/snapshot.storage.k8s.io_volumesnapshotcontents.yaml:48]: According to the doc `If there are no schema changes, the default None conversion strategy may be used and only the apiVersion field will be modified when serving different versions.`.  Since we didn't make any schema changes, I think we don't need to set the `spec.conversion` field and None will be used by default.

File: client/apis/volumesnapshot/v1/types.go
Comment (xing-yang) [reply client/apis/volumesnapshot/v1/types.go:2]: Sure

File: client/hack/README.md
Comment (xing-yang) [reply client/hack/README.md:102]: Yes, we also have doc for the "oneOf".

**Comment by xing-yang** [comment]
@yuxiangqian addressed your comments.  PTAL.

File: client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml
Comment (msau42) [inline client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml:8]: this also has to be updated in all the files

Comment (xing-yang) [reply client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml:8]: You meant I should update the api-approved to the current PR 419?
`api-approved.kubernetes.io: "https://github.com/kubernetes-csi/external-snapshotter/pull/419"`?

Sure.  That makes sense.

Comment (xing-yang) [reply client/config/crd/snapshot.storage.k8s.io_volumesnapshotclasses.yaml]: Done

File: client/apis/volumesnapshot/v1/types.go
Comment (prafull01) [inline client/apis/volumesnapshot/v1/types.go:246]: Currently, the storage and served version `true` is set to the v1beta1. v1 should have `served: true` and `storage: true` and v1beta1 should only have `served: true`. To do that
```suggestion
// +kubebuilder:subresource:status
// +kubebuilder:storageversion
```

File: vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go
Comment (prafull01) [inline vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:246]: Same as above, the storage version should be set to v1

File: vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go
Comment (prafull01) [inline vendor/github.com/kubernetes-csi/external-snapshotter/client/v3/apis/volumesnapshot/v1/types.go:193]: I am not sure if its copied from the above types.go, So those change should be here as well. As suggested in previous comments

File: client/apis/volumesnapshot/v1/types.go
Comment (prafull01) [inline client/apis/volumesnapshot/v1/types.go:194]: Currently, the storage and served version true is set to the v1beta1. v1 should have `served: true` and `storage: true` and v1beta1 should only have served: true. To do that
```suggestion
// +kubebuilder:resource:scope=Cluster
// +kubebuilder:storageversion
```

File: client/apis/volumesnapshot/v1/types.go
Comment (prafull01) [inline client/apis/volumesnapshot/v1/types.go:33]: Currently, the storage and served version true is set to the v1beta1. v1 should have `served: true` and `storage: true` and v1beta1 should only have `served: true`. To do that
```suggestion
// +kubebuilder:subresource:status
// +kubebuilder:storageversion
```

Comment (xing-yang) [reply client/apis/volumesnapshot/v1/types.go:33]: Thanks @prafull01 for checking this!  We actually have to do this in steps.  In 1.20, we move API to v1 but keep storage version at v1beta1.  In 1.21, we plan to change storage version to v1.  This is because we added the validation webhook in 1.19.  See the following KEP for details.
https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/1900-volume-snapshot-validation-webhook

Comment (prafull01) [reply client/apis/volumesnapshot/v1/types.go:33]: Thanks @xing-yang for letting me know. I will remove my comments. :+1:

**Comment by yuxiangqian** [comment]
/lgtm
/hold
for @saad-ali

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:90]:
```suggestion
	// VolumeSnapshotClassName is the name of the VolumeSnapshotClass
	// requested by the VolumeSnapshot.
	// VolumeSnapshotClassName may be left nil to indicate that the default
	// SnapshotClass should be used.
	// A given cluster may have multiple default Volume SnapshotClasses: one
	// default per CSI Driver. If a VolumeSnapshot does not specify a SnapshotClass,
	// VolumeSnapshotSource will be checked to figure out what the associated
	// CSI Driver is, and the default VolumeSnapshotClass associated with that
	// CSI Driver will be used. If more than one VolumeSnapshotClass exist for
	// a given CSI Driver and more than one have been marked as default,
	// CreateSnapshot will fail and generate an event.
	// Empty string is not allowed for this field.
	// +optional
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:104]:
```suggestion
	// This field should be set if the snapshot does not exists, and needs to be
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:112]:
```suggestion
	// This field should be set if the snapshot already exists and only needs a representation in Kubernetes.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:147]: nit:
```suggestion
	// If not specified, it may indicate that the creation time of the snapshot is unknown.
```

Could also mean controller just hasn't gotten around to updating it yet.

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:303]:
```suggestion
	// For dynamically provisioned snapshots, this field will automatically be filled in by the
	// CSI snapshotter sidecar with the "DeletionPolicy" field defined in the corresponding
	// VolumeSnapshotClass.
	// For pre-existing snapshots, users MUST specify this field when creating the
	//  VolumeSnapshotContent object.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:314]:
```suggestion
	// name of the VolumeSnapshotClass from which this snapshot was (or will be)
	// created.
	// Note that after provisioning, the VolumeSnapshotClass may be deleted or
	// recreated with different set of values, and as such, should not be referenced
	// post-snapshot creation.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:318]:
```suggestion
	// source specifies whether the snapshot is (or should be) dynamically provisioned
	// or already exists, and just requires a Kubernetes object representation.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:338]:
```suggestion
	// snapshotHandle specifies the CSI "snapshot_id" of a pre-existing snapshot on
	// the underlying storage system for which a Kubernetes object representation
	// was (or should be) created.
	// This field is immutable.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:371]: `If not specified...`?

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [inline client/apis/volumesnapshot/v1/types.go:399]:
```suggestion
	// error is the last observed error during snapshot creation, if any.
	// Upon success after retry, this error field will be cleared.
```

File: client/apis/volumesnapshot/v1/types.go
Comment (xing-yang) [reply client/apis/volumesnapshot/v1/types.go:107]: Done

**Comment by xing-yang** [comment]
@saad-ali addressed your comments.  Can you please take another look?  Thanks.

File: client/apis/volumesnapshot/v1/types.go
Comment (saad-ali) [reply client/apis/volumesnapshot/v1/types.go:381]: Missed that, disregard.

**Review by saad-ali — APPROVED** [review]
/lgtm
/approve

**Comment by xing-yang** [comment]
@saad-ali has given lgtm and approved this PR.  Removing the hold.

/hold cancel

**Pull request merged**
